{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import gzip\n",
    "import sys\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "from Bio import SeqIO\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "def combine_virus_data(\n",
    "    genomad_summary: Path,\n",
    "    quality_summary: Path,\n",
    "    tantan: Path,\n",
    "    nucleotide_stats: Path,\n",
    ") -> Path:\n",
    "    \"\"\"\n",
    "    Combine virus data from multiple sources into a single TSV file.\n",
    "\n",
    "    Args:\n",
    "        genomad_summary (Path)      : Path to input TSV file containing geNomad summary file.\n",
    "        quality_summary (Path)      : Path to input TSV file containing CheckV's quality_summary.\n",
    "        tantan (Path)               : Path to input TSV file containing tantan's output.\n",
    "        nucleotide_stats (Path)     : Path to input TSV file containing nucleotide stats (GC %, CDS density, and 'N' percentage).\n",
    "\n",
    "    Returns:\n",
    "        Path: Path to output TSV file containing all pertinent virus data combined across sources.\n",
    "    \"\"\"\n",
    "    # load genomad virus summary data\n",
    "    if os.stat(genomad_summary).st_size > 0:\n",
    "        genomad_summary_df = pd.read_csv(\n",
    "            genomad_summary,\n",
    "            sep=\"\\t\",\n",
    "            header=0,\n",
    "            index_col=\"seq_name\",\n",
    "            usecols=[\"seq_name\", \"topology\", \"virus_score\", \"fdr\", \"taxonomy\"]\n",
    "        )\n",
    "    else:\n",
    "        genomad_summary_df = pd.DataFrame(columns=[\"topology\", \"virus_score\", \"fdr\", \"taxonomy\"])\n",
    "\n",
    "    # load checkv quality summary data\n",
    "    if os.stat(quality_summary).st_size > 0:\n",
    "        quality_summary_df = pd.read_csv(\n",
    "            quality_summary,\n",
    "            sep=\"\\t\",\n",
    "            header=0,\n",
    "            index_col=\"contig_id\",\n",
    "            usecols=[\"contig_id\", \"contig_length\", \"provirus\", \"completeness\", \"completeness_method\", \"kmer_freq\", \"warnings\"]\n",
    "        )\n",
    "    else:\n",
    "        quality_summary_df = pd.DataFrame(columns=[\"contig_length\", \"provirus\", \"completeness\", \"completeness_method\", \"kmer_freq\", \"warnings\"])\n",
    "\n",
    "    # load tantan data\n",
    "    if os.stat(tantan).st_size > 0:\n",
    "        tantan_dict = dict([[contig_id, 0] for contig_id in genomad_summary_df.index])\n",
    "        for line in open(tantan):\n",
    "            contig_id, start, end = line.split()\n",
    "            length = int(end) - int(start) + 1\n",
    "            tantan_dict[contig_id] += length\n",
    "        tantan_df = pd.DataFrame.from_dict(tantan_dict, orient=\"index\")\n",
    "        tantan_df.columns = [\"tantan_len\"]\n",
    "    else:\n",
    "        tantan_df = pd.DataFrame(columns=[\"tantan_len\"])\n",
    "\n",
    "    # load nucleotide stats\n",
    "    nucleotide_stats_df = pd.read_csv(\n",
    "        nucleotide_stats,\n",
    "        sep=\"\\t\",\n",
    "        header=0,\n",
    "        index_col=\"contig_id\",\n",
    "        usecols=[\"contig_id\", \"cds_length\", \"cds_density\", \"gc_content\", \"n_count\"]\n",
    "    )\n",
    "    # combine quality data from all sources by contig_id (index)\n",
    "    comb_virus_data_df = pd.concat([\n",
    "        genomad_summary_df,\n",
    "        quality_summary_df,\n",
    "        tantan_df,\n",
    "        nucleotide_stats_df\n",
    "    ], axis=1)\n",
    "\n",
    "\n",
    "    comb_virus_data_df.insert(11, \"tantan_perc\", (comb_virus_data_df['tantan_len'] / comb_virus_data_df['contig_length']))\n",
    "    comb_virus_data_df.insert(len(comb_virus_data_df.columns), \"n_perc\", (comb_virus_data_df['n_count'] / comb_virus_data_df['contig_length']))\n",
    "\n",
    "    return comb_virus_data_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_sequences(\n",
    "    input_fasta: Path,\n",
    "    contigs_to_keep: Path,\n",
    "    comb_virus_data_df: pd.DataFrame,\n",
    "    min_genomad_score: float,\n",
    "    max_genomad_fdr: float,\n",
    "    min_length: int,\n",
    "    min_aai_completeness: float,\n",
    "    min_hmm_completeness: float,\n",
    "    keep_proviruses: bool,\n",
    "    ignore_warnings: bool,\n",
    "    min_cds_density: float,\n",
    "    max_tantan_freq: float,\n",
    "    max_kmer_freq: float,\n",
    "    max_n_freq: float,\n",
    "    output_fasta: Path,\n",
    "    output_tsv: Path\n",
    ") -> None:\n",
    "    \"\"\"\n",
    "    Filter viral sequences based on classification and write to output FASTA/TSV files.\n",
    "\n",
    "    Args:\n",
    "        input_fasta (Path)              : Path to input FASTA file containing terminal repeat sequences.\n",
    "        contigs_to_keep (Path)          : Path to input TSV file (no header) containing contigs to keep (1 contig id per row), regardless of whether they pass filters.\n",
    "        comb_virus_data_df (DataFrame)  : DataFrame containing all pertinent virus data combined across sources.\n",
    "        min_genomad_score (float)      : Minimum geNomad viral score to classify a sequence as viral.\n",
    "        max_genomad_fdr (float)         : Maximum geNomad FDR to classify a sequence as viral.\n",
    "        min_length (int)                : Minimum length to keep a sequence.\n",
    "        keep_proviruses (bool)          : Whether to keep proviruses.\n",
    "        ignore_warnings (bool)          : Whether to ignore warnings.\n",
    "        min_aai_completeness (float)    : Minimum AAI completeness to keep a sequence.\n",
    "        min_hmm_completeness (float)    : Minimum HMM completeness to keep a sequence.\n",
    "        min_cds_density (float)         : Minimum coding density to keep a sequence.\n",
    "        max_tantan_freq (float)         : Maximum tantan frequency to keep a sequence.\n",
    "        max_kmer_freq (float)           : Maximum kmer frequency to keep a sequence.\n",
    "        max_n_freq (float)              : Maximum 'N' frequency to keep a sequence.\n",
    "        output_fasta (Path)             : Path to output FASTA file containing filtered viral sequences.\n",
    "        output_tsv (Path)               : Path to output TSV file containing filtered viral data.\n",
    "\n",
    "    Returns:\n",
    "        Outputs a TSV file with combined virus data, and a FASTA file containing the filtered viral sequences.\n",
    "    \"\"\"\n",
    "    # load contigs to keep\n",
    "    if os.path.exists(str(contigs_to_keep)):\n",
    "        contigs_to_keep_df = pd.read_csv(\n",
    "            contigs_to_keep,\n",
    "            sep=\"\\t\",\n",
    "            header=None,\n",
    "            names=[\"contig_name\"],\n",
    "        )\n",
    "        contigs_to_keep_set = set(contigs_to_keep_df[\"contig_name\"])\n",
    "    else:\n",
    "        contigs_to_keep_set = set()\n",
    "\n",
    "    if not keep_proviruses:\n",
    "        comb_virus_data_df = comb_virus_data_df.query(\"`provirus` == 'No' and `topology` != 'Provirus'\")\n",
    "    \n",
    "    if not ignore_warnings:\n",
    "        comb_virus_data_df = comb_virus_data_df.query(\"`warnings` != 'contig >1.5x longer than expected genome length'\")\n",
    "\n",
    "    # filter viral sequences by classification, completeness, and quality\n",
    "    is_virus_filter = \"(`virus_score` >= @min_genomad_score and `fdr` <= @max_genomad_fdr)\"\n",
    "    is_hq_filter = \"(`completeness` >= @min_aai_completeness or `completeness` >= @min_hmm_completeness or `contig_length` >= @min_length)\"\n",
    "    seq_filter = \"(`cds_density` >= @min_cds_density and `tantan_perc` <= @max_tantan_freq and `kmer_freq` <= @max_kmer_freq and `n_perc` <= @max_n_freq)\"\n",
    "    contigs_passing_filters_set = set(\n",
    "        comb_virus_data_df.query(\n",
    "            f\"{is_virus_filter} and {is_hq_filter} and {seq_filter}\"\n",
    "        ).index\n",
    "    )\n",
    "\n",
    "    filtered_contigs_set = contigs_to_keep_set.union(contigs_passing_filters_set)\n",
    "    # write filtered viral sequences to output FASTA\n",
    "    with gzip.open(input_fasta, \"rt\") as input_gunzipped, open(output_fasta, \"w\") as output:\n",
    "        for r in SeqIO.parse(input_gunzipped, \"fasta\"):\n",
    "            if r.id in filtered_contigs_set:\n",
    "                SeqIO.write(r, output, \"fasta\")\n",
    "    # write combined metadata to output TSV\n",
    "    comb_virus_data_df.insert(0, \"contig_id\", comb_virus_data_df.index)\n",
    "    comb_virus_data_df.to_csv(output_tsv, sep=\"\\t\", index=False)\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_data_df = combine_virus_data(\n",
    "    genomad_summary = \"/mmfs1/gscratch/pedslabs_hoffman/carsonjm/CarsonJM/HoffLab-phageannotator/results/VirusClassification/genomad/endtoend/group-group1_virus_summary.tsv\",\n",
    "    quality_summary = \"/mmfs1/gscratch/pedslabs_hoffman/carsonjm/CarsonJM/HoffLab-phageannotator/results/VirusQuality/checkv/endtoend/group-group1_quality_summary.tsv\",\n",
    "    nucleotide_stats = \"/mmfs1/gscratch/pedslabs_hoffman/carsonjm/CarsonJM/HoffLab-phageannotator/results/VirusQuality/nucleotidestats/group-group1_nuc_stats.tsv\",\n",
    "    tantan = \"/mmfs1/gscratch/pedslabs_hoffman/carsonjm/CarsonJM/HoffLab-phageannotator/results/VirusQuality/tantan/group-group1_tantan.tsv\"\n",
    ")\n",
    "\n",
    "\n",
    "filter_sequences(\n",
    "    input_fasta = \"/mmfs1/gscratch/pedslabs_hoffman/carsonjm/CarsonJM/HoffLab-phageannotator/results/Assembly/metaspades/coassembly/group-group1.scaffolds.fa.gz\",\n",
    "    comb_virus_data_df = combined_data_df,\n",
    "    contigs_to_keep = \"NULL\",\n",
    "    min_genomad_score = 0.7,\n",
    "    max_genomad_fdr = 0.1,\n",
    "    min_length = 100000,\n",
    "    keep_proviruses = False,\n",
    "    ignore_warnings = False,\n",
    "    min_aai_completeness = 50,\n",
    "    min_hmm_completeness = 50,\n",
    "    min_cds_density = 70,\n",
    "    max_tantan_freq = 0.05,\n",
    "    max_kmer_freq = 1.15,\n",
    "    max_n_freq = 0.01,\n",
    "    output_fasta = \"/mmfs1/gscratch/pedslabs_hoffman/carsonjm/CarsonJM/HoffLab-phageannotator/results/test.fasta\",\n",
    "    output_tsv = \"/mmfs1/gscratch/pedslabs_hoffman/carsonjm/CarsonJM/HoffLab-phageannotator/results/test.tsv\"\n",
    ")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "2.7.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
